Add Vote Extension varying size testnet (#888)
cometbft/cometbft: docs/qa/CometBFT-QA-38.md

### CPU

The CPU usage reached the same peaks on all tests, but the following graphs show that with larger Vote Extensions, nodes take longer to reduce the CPU usage.
This could mean that a backlog of processing is forming during the execution of the tests with larger extensions.

| Experiment   | CPU                                                   |
| ------------ | ----------------------------------------------------- |
| **baseline** | ![cpu-avg](img38/voteExtensions/baseline_avg_cpu.png) |
| **2k**       | ![cpu-avg](img38/voteExtensions/02k_avg_cpu.png)      |
| **4k**       | ![cpu-avg](img38/voteExtensions/04k_avg_cpu.png)      |
| **8k**       | ![cpu-avg](img38/voteExtensions/08k_avg_cpu.png)      |
| **16k**      | ![cpu-avg](img38/voteExtensions/16k_avg_cpu.png)      |
| **32k**      | ![cpu-avg](img38/voteExtensions/32k_avg_cpu.png)      |

### Resident Memory

The same conclusion reached for CPU usage may be drawn for the memory.
That is, that a backlog of work is formed during the tests and catching up (freeing of memory) happens after the test is done.

> :warning:

> > A more worrying trend is that the bottom of the memory usage seems to increase in between runs.
> > This will be investigated in a future issue.
> 

Author: These images show the memory usage over 5 twenty minute runs.
They don't indicate any up trend in memory consumption.

min ![image](https://github.com/cometbft/cometbft/assets/2686194/d6824506-0002-45b3-8805-b05e11d53a36)

max  ![image](https://github.com/cometbft/cometbft/assets/2686194/286c2c14-7ffc-41ea-b88d-7bcce4db926f)

avg 
![image](https://github.com/cometbft/cometbft/assets/2686194/03fbeee3-40a7-4caf-857c-df3edf986c08)

actual values 
![image](https://github.com/cometbft/cometbft/assets/2686194/0964721d-6797-4e3e-a0ad-e180e420da45)

```suggestion
A more worrying trend is that the bottom of the memory usage seems to increase in between runs.
We have investigated this in longer runs and confirmed that there is no such a trend.

```

### Mempool size

This metric shows how many transactions are outstanding in the nodes' mempools.
Observe that in all runs, the average number of transactions in the mempool quickly drops to near zero between runs.

| Experiment   | Resident Set Size                                                  |
| ------------ | ------------------------------------------------------------------ |
| **baseline** | ![mempool-avg](img38/voteExtensions/baseline_avg_mempool_size.png) |
| **2k**       | ![mempool-avg](img38/voteExtensions/02k_avg_mempool_size.png)      |
| **4k**       | ![mempool-avg](img38/voteExtensions/04k_avg_mempool_size.png)      |
| **8k**       | ![mempool-avg](img38/voteExtensions/08k_avg_mempool_size.png)      |
| **16k**      | ![mempool-avg](img38/voteExtensions/16k_avg_mempool_size.png)      |
| **32k**      | ![mempool-avg](img38/voteExtensions/32k_avg_mempool_size.png)      |

### Results

| Scenario | Date       | Version                                                                               | Result |
| -------- | ---------- | ------------------------------------------------------------------------------------- | ------ |

> | VESize | 2023-05-23 | v0.38.0-alpha.2 + varying vote extensions  (9fc711b6514f99b2dc0864fc703cb81214f01783) |        |

Reviewer: ```suggestion
| VESize   | 2023-05-23 | v0.38.0-alpha.2 + varying vote extensions  (9fc711b6514f99b2dc0864fc703cb81214f01783) |  PASS  |
```

I guess you have earned the right to fill that in with "pass" :-)

Reviewer: Sorry, after offline discussion with @lasarojc I realise, it should be "N/A" because this is the baseline for future releases (there's nothing in the past to compare this to)

We conjecture that the reason is that the timeouts used are inadequate for the extra traffic in the network.

### CPU

The CPU usage reached the same peaks on all tests, but the following graphs show that with larger Vote Extensions, nodes take longer to reduce the CPU usage.
This could mean that a backlog of processing is forming during the execution of the tests with larger extensions.

| Experiment   | CPU                                                   |
| ------------ | ----------------------------------------------------- |
| **baseline** | ![cpu-avg](img38/voteExtensions/baseline_avg_cpu.png) |
| **2k**       | ![cpu-avg](img38/voteExtensions/02k_avg_cpu.png)      |
| **4k**       | ![cpu-avg](img38/voteExtensions/04k_avg_cpu.png)      |
| **8k**       | ![cpu-avg](img38/voteExtensions/08k_avg_cpu.png)      |
| **16k**      | ![cpu-avg](img38/voteExtensions/16k_avg_cpu.png)      |
| **32k**      | ![cpu-avg](img38/voteExtensions/32k_avg_cpu.png)      |

### Resident Memory

The same conclusion reached for CPU usage may be drawn for the memory.
That is, that a backlog of work is formed during the tests and catching up (freeing of memory) happens after the test is done.

> > :warning:

Author: ```suggestion
```

| **baseline** | ![block rate](img38/voteExtensions/baseline_1_block_rate.png) | ![txs rate](img38/voteExtensions/baseline_1_total_txs_rate.png) |
| **2k**       | ![block rate](img38/voteExtensions/02k_1_block_rate.png)      | ![txs rate](img38/voteExtensions/02k_1_total_txs_rate.png)      |
| **4k**       | ![block rate](img38/voteExtensions/04k_1_block_rate.png)      | ![txs rate](img38/voteExtensions/04k_1_total_txs_rate.png)      |
| **8k**       | ![block rate](img38/voteExtensions/08k_1_block_rate.png)      | ![txs rate](img38/voteExtensions/08k_1_total_txs_rate.png)      |
| **16k**      | ![block rate](img38/voteExtensions/16k_1_block_rate.png)      | ![txs rate](img38/voteExtensions/16k_1_total_txs_rate.png)      |
| **32k**      | ![block rate](img38/voteExtensions/32k_1_block_rate.png)      | ![txs rate](img38/voteExtensions/32k_1_total_txs_rate.png)      |

### Number of rounds

The effect of vote extensions are also felt on the number of rounds needed to reach consensus.
The following graphs show the number of the highest round required to reach consensus during the whole experiment.

> In the baseline and low loads, most blocks were agreed upon during round 0.

Reviewer: ```suggestion
In the baseline and low vote extension lengths, most blocks were agreed upon during round 0.
```

> As the load increases, more and more blocks were required.

Reviewer: ```suggestion
As the load increases, more and more rounds were required.
```

| Experiment   | CPU                                                   |
| ------------ | ----------------------------------------------------- |
| **baseline** | ![cpu-avg](img38/voteExtensions/baseline_avg_cpu.png) |
| **2k**       | ![cpu-avg](img38/voteExtensions/02k_avg_cpu.png)      |
| **4k**       | ![cpu-avg](img38/voteExtensions/04k_avg_cpu.png)      |
| **8k**       | ![cpu-avg](img38/voteExtensions/08k_avg_cpu.png)      |
| **16k**      | ![cpu-avg](img38/voteExtensions/16k_avg_cpu.png)      |
| **32k**      | ![cpu-avg](img38/voteExtensions/32k_avg_cpu.png)      |

### Resident Memory

The same conclusion reached for CPU usage may be drawn for the memory.
That is, that a backlog of work is formed during the tests and catching up (freeing of memory) happens after the test is done.

> :warning:
> A more worrying trend is that the bottom of the memory usage seems to increase in between runs.
> This will be investigated in a future issue.

| Experiment   | Resident Set Size                                        |
| ------------ | -------------------------------------------------------- |
| **baseline** | ![rss-avg](img38/voteExtensions/baseline_avg_memory.png) |

> | **8k**       | ![rss-avg](img38/voteExtensions/02k_avg_memory.png)      |

Reviewer: ```suggestion
| **2k**       | ![rss-avg](img38/voteExtensions/02k_avg_memory.png)      |
```

> | **8k**       | ![rss-avg](img38/voteExtensions/04k_avg_memory.png)      |

Reviewer: ```suggestion
| **4k**       | ![rss-avg](img38/voteExtensions/04k_avg_memory.png)      |
```
I imagine the graphs correspond to the VE sizes

| **8k**       | ![rss-avg](img38/voteExtensions/08k_avg_memory.png)      |
| **16k**      | ![rss-avg](img38/voteExtensions/16k_avg_memory.png)      |
| **32k**      | ![rss-avg](img38/voteExtensions/32k_avg_memory.png)      |

### Mempool size

This metric shows how many transactions are outstanding in the nodes' mempools.
Observe that in all runs, the average number of transactions in the mempool quickly drops to near zero between runs.

| Experiment   | Resident Set Size                                                  |
| ------------ | ------------------------------------------------------------------ |
| **baseline** | ![mempool-avg](img38/voteExtensions/baseline_avg_mempool_size.png) |

> | **8k**       | ![mempool-avg](img38/voteExtensions/02k_avg_mempool_size.png)      |

Reviewer: ```suggestion
| **2k**       | ![mempool-avg](img38/voteExtensions/02k_avg_mempool_size.png)      |
```

> | **8k**       | ![mempool-avg](img38/voteExtensions/04k_avg_mempool_size.png)      |

Reviewer: ```suggestion
| **4k**       | ![mempool-avg](img38/voteExtensions/04k_avg_mempool_size.png)      |
```

| **8k**       | ![mempool-avg](img38/voteExtensions/08k_avg_mempool_size.png)      |
| **16k**      | ![mempool-avg](img38/voteExtensions/16k_avg_mempool_size.png)      |
| **32k**      | ![mempool-avg](img38/voteExtensions/32k_avg_mempool_size.png)      |

#### Observations/Questions/TODOs

* what is the effect on block sizes
    * how is the block size limited? do the extensions fit? the proposal itself is small.
    * block sizes depending on the extension sizes
* what is the effect on the network usage
* txs per block depending on the extension sizes
    * remember that there is one special transaction
* we need to evaluate the effect of transaction sizes as well.
* the mempool never got completely empty on some nodes
* 5 minute interval to clean up the mempool. heights continue to happen, with vote extensions.
* if the mempool size drops quickly, why is the memory slow to empty? Slow GC?
* run with longer timeouts

> * run baseline for 20 minutes straight.

Reviewer: Could we move this text to a (tracking) issue?

Author: Done 903.

### Results

| Scenario | Date       | Version                                                                               | Result |
| -------- | ---------- | ------------------------------------------------------------------------------------- | ------ |

> | Rotating | 2023-05-23 | v0.38.0-alpha.2 + varying vote extensions  (9fc711b6514f99b2dc0864fc703cb81214f01783) |        |

Reviewer: Also: "Result = PASS" ?

#### Conclusions

[\#9539]: https://github.com/tendermint/tendermint/issues/9539
[\#9548]: https://github.com/tendermint/tendermint/issues/9548
[\#539]: https://github.com/cometbft/cometbft/issues/539
[\#546]: https://github.com/cometbft/cometbft/issues/546
[\#562]: https://github.com/cometbft/cometbft/issues/562

> [end-to-end]: TODO

Reviewer: ?

Author: Fixed.
